### PR TITLE
Introduce Kotlin MPP to multicast module

### PR DESCRIPTION
**Pull request by tasomaniac**
## Considerations

- It is jvm-only for now.
- Should have no behavior change
- Jacoco test report and apiDump works as before

### TODO

- [ ] Maven publish needs to be figured out

### CODE REVIEW

File: multicast/src/commonMain/kotlin/com/dropbox/flow/multicast/StoreRealActor.kt [hunk 1]
```diff
@@ -0,0 +1,12 @@
+package com.dropbox.flow.multicast
+
+import kotlinx.coroutines.CoroutineScope
+
+internal expect abstract class StoreRealActor<T>(
```
Comment (clhols) [inline multicast/src/commonMain/kotlin/com/dropbox/flow/multicast/StoreRealActor.kt:5]: Why use expect/actual on this class?

Comment (digitalbuddha) [reply multicast/src/commonMain/kotlin/com/dropbox/flow/multicast/StoreRealActor.kt:5]: Currently it is backed by kotlin actor which is not compatible with native afaik

**Comment by clhols** [comment]
> You can use the atomicfu lib instead of an expect/actual of AtomicBoolean.

Now I remember having tried to replace AtomicBoolean with  atomicfu earlier, but jacoco didn't like it.
Gave me this when running "check":
```
Caused by: java.io.IOException: Error while analyzing StoreRealActor.class.
        at org.jacoco.core.analysis.Analyzer.analyzerError(Analyzer.java:162)
        at org.jacoco.core.analysis.Analyzer.analyzeClass(Analyzer.java:134)
        at org.jacoco.core.analysis.Analyzer.analyzeClass(Analyzer.java:157)
        at org.jacoco.core.analysis.Analyzer.analyzeAll(Analyzer.java:193)
        at org.jacoco.ant.ReportTask.createBundle(ReportTask.java:573)
        at org.jacoco.ant.ReportTask.createReport(ReportTask.java:545)
        at org.jacoco.ant.ReportTask.execute(ReportTask.java:496)
        ... 117 more
Caused by: java.lang.IllegalStateException: Can't add different class with same name: com/dropbox/flow/multicast/StoreRealActor
        at org.jacoco.core.analysis.CoverageBuilder.visitCoverage(CoverageBuilder.java:106)
        at org.jacoco.core.analysis.Analyzer$1.visitEnd(Analyzer.java:99)
        at org.objectweb.asm.ClassVisitor.visitEnd(ClassVisitor.java:326)
        at org.jacoco.core.internal.flow.ClassProbesAdapter.visitEnd(ClassProbesAdapter.java:100)
        at org.objectweb.asm.ClassReader.accept(ClassReader.java:692)
        at org.objectweb.asm.ClassReader.accept(ClassReader.java:400)
        at org.jacoco.core.analysis.Analyzer.analyzeClass(Analyzer.java:116)
        at org.jacoco.core.analysis.Analyzer.analyzeClass(Analyzer.java:132)
```
So I guess expect/actual on AtomicBoolean is better for now.

File: multicast/src/commonMain/kotlin/com/dropbox/flow/multicast/StoreRealActor.kt (hunk 1 above)
Comment (clhols) [reply multicast/src/commonMain/kotlin/com/dropbox/flow/multicast/StoreRealActor.kt:5]: Correct, now I remember.

**Comment by tasomaniac** [comment]
I would also prefer to merge it as is to keep the behavior as before for jvm. Once we have more modules with KMP plugin, then we can migrate them to atomic-fu. Let me know if that makes sense

**Comment by clhols** [comment]
> I would also prefer to merge it as is to keep the behavior as before for jvm. Once we have more modules with KMP plugin, then we can migrate them to atomic-fu. Let me know if that makes sense

It depends on when we want to add native targets and if it is wise to implement our own native version of AtomicBoolean to do that.
I did manage to fix the atomicfu issue by adding '**/atomicfu' to the jacoco exclude filter.

**Comment by tasomaniac** [comment]
@clhols thanks for the hints. The branch is rebased. Now `atomicfu` and `kotlin.test` are used.

File: buildsystem/dependencies.gradle [hunk 2]
```diff
@@ -66,7 +66,9 @@ ext.libraries = [
         truth                 : "com.google.truth:truth:$versions.truth",
         mockito               : "org.mockito:mockito-core:$versions.mockito",
         mockitoKotlin         : "com.nhaarman.mockitokotlin2:mockito-kotlin:$versions.mockitoKotlin",
+        coroutinesCommon      : "org.jetbrains.kotlinx:kotlinx-coroutines-core-common:$versions.coroutines",
         coroutinesCore        : "org.jetbrains.kotlinx:kotlinx-coroutines-core:$versions.coroutines",
+        coroutinesNative        : "org.jetbrains.kotlinx:kotlinx-coroutines-core-native:$versions.coroutines",
```
Comment (pavlospt) [inline buildsystem/dependencies.gradle:71]: @tasomaniac maybe a formatting issue here 😄

**Comment by clhols** [comment]
> @clhols thanks for the hints. The branch is rebased. Now `atomicfu` and `kotlin.test` are used.

Yeah, I needed atomicfu for the Cache module and its tests also ;)

**Comment by clhols** [comment]
You should be able to put the tests in commonTest now that they don't rely on junit anymore.

**Comment by yigit** [comment]
Can we move this work to a branch instead?
it is not like we'll benefit from KN right now since we heavily rely on coroutines which barely work there.
It would be better not to take more instability so that we can focus on shipping.

**Comment by tasomaniac** [comment]
I picked it up because from the open issues it seemed like KMP was a priority.

Adding the Kotlin MP plugin is not a big change. Everything can be kept jvm only. Only change would be instead of having main folder, you will have jvmMain.

Otherwise, I don't think it is a good idea to keep it in a separate branch since it will be crazy hard to fix the conflicts (correct me if I'm wrong).

I would suggest either to pause Multiplatform convertion until the stable release if adding the plugin and keeping it jvm only is not an option.

**Comment by yigit** [comment]
this is on us, sorry about it; we didn't discuss fully before jumping into it.
KMP brings a lot of unknown, even the plugin itself is a rapidly moving target. + we cannot really benefit much at this point since coroutines are also very limited in native at this point.

It is important for us to ship a stable version at this point instead of taking more instability. We can start the KMP work once library is stable. It probably makes sense to branch master out once we call it beta.

**Comment by tasomaniac** [comment]
Agreed 👍

**Pull request closed without merging**